### PR TITLE
feat: add hub push cli

### PR DESCRIPTION
**Pull request by numb3r3**
Usage:

```bash
$ jina hub push [--private/--public] [--force] [--secret] [PATH]
```

Now, you can work with local hubble system via http://localhost:3001 (default host is: https://hubble.jina.ai (TBD)):
```bash
$ JINA_HUBBLE_REGISTRY=http://localhost:3001 jina hub push /path/to/executor --public
```

or you can run unit-test directly
```bash
$ pytest -v tests/unit/hubble
```

### CODE REVIEW

File: jina/hubble/helper.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def archive_package(package_folder: 'Path') -> 'io.BytesIO':
+    """
+    Archives the given folder in zip format and return a data stream.
+    :param package_folder: the folder path of the package
+    :return: the data stream of zip content
+    """
+    import os
+    import zipfile
```
Comment (hanxiao) [inline jina/hubble/helper.py:14]: standard python library can be imported at header, as they are usually very fast to load

File: jina/hubble/helper.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    ignored_file_specs = ['**/*.pyc']
+    ignored_dir_specs = [
+        '__MACOSX',
+        '.DS_Store',
+        '__pycache__',
+        '.eggs',
+        '*.egg-info',
+        '.git',
+        '.vscode',
+    ]
```
Comment (hanxiao) [inline jina/hubble/helper.py:28]: how about:
- put this file https://github.com/github/gitignore/blob/master/Python.gitignore under `jina/resources`
- load it via
```python
from jina import __resources_path__
with open(os.path.join(__resources_path__, 'Python.gitignore')) as fp:
   ignored_dir_specs = fp.readlines()
```

File: jina/hubble/hubio.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    def push(self) -> None:
+        """Push the executor pacakge to Jina Hub."""
+
+        import requests
+        import hashlib
+
+        is_public = True
+        if self.args.private:
+            is_public = False
+            self.args.public = False
```
Comment (hanxiao) [inline jina/hubble/hubio.py:44]: this looks redundant, we only need `--public` we don't need `--private`,

when not given `--public` (by default), then `self.args.public` will be False, so it is private.

so default is `private`

File: jina/hubble/hubio.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                'meta': json.dumps(meta),
+                'env': json.dumps(env),
+                'is_public': is_public,
+                'md5sum': md5_digest,
+                'force': self.args.force,
+                'secret': self.args.secret,
+            }
+
+            # upload the archived executor to Jina Hub
+            upload_url = HUBBLE_REGISTRY + '/v1/executors/push'
```
Comment (hanxiao) [inline jina/hubble/hubio.py:77]: put it in the header after `HUBBLE_REGISTRY` for better visibility and maintainbility

File: jina/hubble/hubio.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                    f'There is some errors while pushing executor "{self.args.path}"'
+                )
+
+        except Exception as e:  # IO related errors
+            self.logger.error(
+                f'Error when trying to push the executor at {self.args.path}: {e!r}'
+            )
+            raise e
+
+        self.logger.success(f'🎉 The executor located at {pkg_path} is now published!')
```
Comment (hanxiao) [inline jina/hubble/hubio.py:97]: i think this success msg is wrongly placed as even an error will lead to this line.

File: jina/parsers/hubble/push.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+The content source to be shipped into a Jina Hub executor. It can one of the followings:
+- a directory containing Dockerfile, manifest.yml, README.md, zero or more yaml config, zero or more Python file.
+''',
+    )
+    gp.add_argument(
+        '--private',
+        action='store_true',
+        default=False,
+        help='If set, the published executor is invisible to public',
+    )
```
Comment (hanxiao) [inline jina/parsers/hubble/push.py:23]:
```suggestion

```

File: jina/parsers/hubble/push.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    gp.add_argument(
+        '--public',
+        action='store_true',
+        default=True,
```
Comment (hanxiao) [inline jina/parsers/hubble/push.py:27]:
```suggestion
        default=False,
```

File: jina/parsers/hubble/push.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        help='If set, the published executor is visible to public',
+    )
+    gp.add_argument(
+        '--force',
+        type=str,
+        default='',
```
Comment (hanxiao) [inline jina/parsers/hubble/push.py:33]:
```suggestion

```
no need

File: jina/parsers/hubble/push.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        help='To overwrite the executor identified as UUID8',
+    )
+    gp.add_argument(
+        '--secret',
+        type=str,
+        default='',
```
Comment (hanxiao) [inline jina/parsers/hubble/push.py:39]:
```suggestion
```
no need

**Review by hanxiao — CHANGES_REQUESTED** [review]
please change accordingly

File: jina/hubble/hubio.py (hunk 3 above)
Comment (numb3r3) [reply jina/hubble/hubio.py]: done

**Review by hanxiao — APPROVED** [review]

File: jina/hubble/hubio.py (hunk 3 above)
Comment (mapleeit) [reply jina/hubble/hubio.py]: @hanxiao I think we should keep the `--public` and `--private`. We may use them with the `--force` option.

We people do `jina hub push --public --force UUID8 --secret TOKEN .`. They may want to change the visibility of this executor. From private to public, or from public to private.

Comment (mapleeit) [reply jina/hubble/hubio.py]: And why the default is `private`? As far as I can see, if we want more public shared executors in jina hub, public default value could be better.

**Pull request merged**